improver/psychrometric_calculations/condensation_trails.py

```diff
@@ -367,3 +304,152 @@ def process(self, *cubes: Union[Cube, CubeList]) -> Cube:
         )
 
         return return_cube
+
+
+class ContrailHeightExtractor(BasePlugin):
+    """
+    Plugin to extract contrail formation heights by category. It extracts the maximum or minimum
+    height where contrail formation is Non-persistent or Persistent.
+    """
+
+    def __init__(self, use_max=True):
+        """
+        Initialize the Class
+
+        Args:
+            use_max (bool, optional):
+                If True, extract maximum heights; if False, extract minimum heights.
+        """
+
+        self.use_max = use_max
+
+    def _define_max_min_height_contrail_cubes(
+        self,
+        formation_cube: Cube,
+        height_cube: Cube,
+        non_persistent_result: np.ndarray,
+        persistent_result: np.ndarray,
+        operation: str,
+    ) -> Tuple[Cube, Cube]:
+        """
+        Create new cubes containing the max or min heights
+        for persistent or non-persistent contrail formation.
+
+        Args:
+            formation_cube (iris.cube.Cube):
+                Categorical cube of shape (engine_factor, pressure_level, lat (optional), lon (optional))
+            height_cube (iris.cube.Cube):
+                Height cube of shape (pressure_level, lat (optional), lon (optional))
+            non_persistent_result (np.ndarray):
+                Extracted height data for non-persistent contrails.
+            persistent_result (np.ndarray):
+                Extracted height data for persistent contrails.
+            operation (str):
```

Comment:
```suggestion
            formation_cube:
                Categorical cube of shape (engine_factor, pressure_level, lat (optional), lon (optional))
            height_cube:
                Height cube of shape (pressure_level, lat (optional), lon (optional))
            non_persistent_result:
                Extracted height data for non-persistent contrails.
            persistent_result:
                Extracted height data for persistent contrails.
            operation:
```
Types not required in doc string, as these are included as type hints in the function signature.